voltdrop_direct/index.js

```diff
@@ -0,0 +1,448 @@
+ */
+
+/**
+ * Downlink decode
+ * @param {Object} input - An object provided by the IoT Flow framework
+ * @param {number[]} input.bytes - Array of bytes represented as numbers as it will be sent to the device
+ * @param {number} input.fPort - The Port Field on which the downlink must be sent
+ * @param {Date} input.recvTime - The uplink message time computed by the IoT Flow framework
+ * @returns {DecodedDownlink} The decoded object
+ */
+function decodeDownlink(input) {
+  let result = {
+    data: {},
+    errors: [],
+    warnings: [],
+  };
+
+  let raw = Buffer.from(input.bytes);
+
+  if (raw.length < 2) {
+    result.errors.push("Invalid downlink: downlink must be 2 bytes or greater");
+    delete result.data;
+    return result;
+  }
+
+  let type = raw.readUInt16BE(0);
+  switch (type) {
+    case 0x31: // transmit interval
+      if (raw.length < 6) {
+        result.errors.push(
+          "Invalid downlink: transmit interval downlink must be 6 bytes or greater"
+        );
+        delete result.data;
+        return result;
+      }
+      let transmitIntervalSeconds = raw.readUInt32BE(2);
+      if (transmitIntervalSeconds < 60) {
+        result.errors.push(
+          "Invalid downlink: transmit interval cannot be less than 1 min"
+        );
+        delete result.data;
+        return result;
+      }
+      if (transmitIntervalSeconds > 1800) {
+        result.errors.push(
+          "Invalid downlink: transmit interval cannot be greater than 30 min"
+        );
+        delete result.data;
+        return result;
+      }
+      result.data.transmitIntervalSeconds = transmitIntervalSeconds;
+      break;
+    case 0x30: // packet transmit schedule
+      if (raw.length < 4) {
+        result.errors.push(
+          "Invalid downlink: transmit packet schedule downlink must be 4 bytes or greater"
+        );
+        delete result.data;
+        return result;
+      }
+      let dataLength = raw.readUInt8(2);
+      if (dataLength > (raw.length - 3)) {
+        result.errors.push(
+          "Invalid downlink: transmit packet schedule length field larger than input data"
+        );
+        delete result.data;
+        return result;
+      } else if (dataLength == 0) {
+        result.errors.push(
+          "Invalid downlink: Packet transmit schedule length field cannot be zero"
+        );
+        delete result.data;
+        return result;
+      } else if (dataLength > 60) {
+        result.errors.push(
+          "Invalid downlink: Packet transmit schedule cannot be longer than 60 elements"
+        );
+        delete result.data;
+        return result;
+      }
+      const validIDList = [0, 40, 41, 42, 43, 44, 45];
+      let scheduleValid = false;
+      result.data.packetTransmitSchedule = [];
+      for (let index = 0; index < dataLength; index++) {
+        let id = raw.readUint8(index + 3);
+        if (!validIDList.includes(id)) {
```

Comment:
IMO, we should be saying there is an invalid packet ID, but we shouldn't be replacing it with a 0 because that is modifying the downlink we are trying to decode. Besides any necessary conversions or calculations to interpret data, I don't think we should be modifying the underlying downlink info.
The warnings should be sufficient to indicate to the user that the downlink is invalid.
We could modify the message to say that invalid packets will be replaced with 0 when they are being encoded.
Also see my other comment about including the index of the invalid packet ID.